db/compaction/compaction_job.cc

```diff
@@ -290,6 +294,141 @@ void CompactionJob::Prepare() {
+          max_subcompactions_limit;
+    }
+  } else {
+    num_planned_subcompactions =
+        std::min({static_cast<uint64_t>(ranges_size_limit),
+                  max_subcompactions_limit, max_output_files_limit});
+  }
+  return num_planned_subcompactions;
+}
+
+void CompactionJob::ReserveSubcompactionResources(
+    int extra_required_subcompactions) {
+  // Apply min function first since We need to compute the extra subcompaction
+  // against compaction limits. And then try to reserve threads for extra
+  // subcompactions. The actual number of reserved threads could be less than
+  // the desired number.
+  int max_db_compactions_ =
+      DBImpl::GetBGJobLimits(
+          mutable_db_options_copy_.max_background_flushes,
+          mutable_db_options_copy_.max_background_compactions,
+          mutable_db_options_copy_.max_background_jobs,
+          versions_->GetColumnFamilySet()
+              ->write_controller()
+              ->NeedSpeedupCompaction())
+          .max_compactions;
+  int available_bg_compactions_against_db_limit =
+      std::max(max_db_compactions_ - *bg_compaction_scheduled_ -
+                   *bg_bottom_compaction_scheduled_,
+               0);
+  extra_num_subcompaction_threads_reserved_ =
+      env_->ReserveThreads(std::min(extra_required_subcompactions,
+                                    available_bg_compactions_against_db_limit),
+                           thread_pri_);
+
+  db_mutex_->AssertHeld();
+  // Update bg_compaction_scheduled_ or bg_bottom_compaction_scheduled_
+  // depending on if this compaction has the bottommost priority
+  if (thread_pri_ == Env::Priority::BOTTOM && bottommost_level_) {
+    if (bg_bottom_compaction_scheduled_ != nullptr) {
+      *bg_bottom_compaction_scheduled_ +=
+          extra_num_subcompaction_threads_reserved_;
+    }
+  } else if (bg_compaction_scheduled_ != nullptr) {
+    *bg_compaction_scheduled_ += extra_num_subcompaction_threads_reserved_;
+  }
+}
+
+void CompactionJob::ShrinkSubcompactionResources(
+    uint64_t num_unnecessary_subcompactions) {
+  // We cannot release threads more than what we reserved before
+  int extra_num_subcompaction_threads_released =
+      env_->ReleaseThreads(std::min((int)num_unnecessary_subcompactions,
+                                    extra_num_subcompaction_threads_reserved_),
+                           thread_pri_);
+  // Update the number of reserved threads and the number of background
+  // scheduled compactions for this compaction job
+  extra_num_subcompaction_threads_reserved_ -=
+      extra_num_subcompaction_threads_released;
+  db_mutex_->AssertHeld();
+  // Update bg_compaction_scheduled_ or bg_bottom_compaction_scheduled_
+  // depending on if this compaction has the bottommost priority
+  if (thread_pri_ == Env::Priority::BOTTOM && bottommost_level_) {
+    if (bg_bottom_compaction_scheduled_ != nullptr) {
+      *bg_bottom_compaction_scheduled_ -= (int)num_unnecessary_subcompactions;
```

Comment:
Should this be `extra_num_subcompaction_threads_released`? Same as below.

Reply:
Nice catch!